server/base/src/main/java/org/apache/accumulo/server/util/UpgradePreparationUtil.java

```diff
@@ -0,0 +1,155 @@
+    return "utility used to prepare an instance for a minor or major version upgrade";
+  }
+
+  @Override
+  public void execute(String[] args) throws Exception {
+    Opts opts = new Opts();
+    opts.parseArgs(keyword(), args);
+
+    if (!opts.postShutdownUpgradeCheck) {
+      new JCommander(opts).usage();
+      return;
+    }
+
+    var siteConf = SiteConfiguration.auto();
+    // Login as the server on secure HDFS
+    if (siteConf.getBoolean(Property.INSTANCE_RPC_SASL_ENABLED)) {
+      SecurityUtil.serverLogin(siteConf);
+    }
+
+    String volDir = VolumeConfiguration.getVolumeUris(siteConf).iterator().next();
+    Path instanceDir = new Path(volDir, "instance_id");
+    InstanceId iid = VolumeManager.getInstanceIDFromHdfs(instanceDir, new Configuration());
+    ZooReaderWriter zoo = new ZooReaderWriter(siteConf);
+
+    if (opts.postShutdownUpgradeCheck) {
+      final String zUpgradepath = Constants.ZROOT + "/" + iid + Constants.ZPREPARE_FOR_UPGRADE;
+      try {
+        if (zoo.exists(zUpgradepath)) {
+          if (!opts.force) {
+            throw new IllegalStateException(
+                "'accumulo upgrade --prepare-for-upgrade' must have already been run."
+                    + " To run again use 'accumulo upgrade --prepare-for-upgrade --force'");
+          } else {
+            zoo.delete(zUpgradepath);
+          }
+        }
+      } catch (KeeperException | InterruptedException e) {
+        throw new IllegalStateException("Error creating or checking for " + zUpgradepath
+            + " node in zookeeper: " + e.getMessage(), e);
+      }
+
+      LOG.info("Upgrade specified, validating that Manager is stopped");
+      final ServiceLockPath mgrPath =
+          ServiceLock.path(Constants.ZROOT + "/" + iid + Constants.ZMANAGER_LOCK);
+      try {
+        if (ServiceLock.getLockData(zoo.getZooKeeper(), mgrPath) != null) {
+          throw new IllegalStateException(
+              "Manager is running, shut it down and retry this operation");
+        }
+      } catch (KeeperException | InterruptedException e) {
+        throw new IllegalStateException("Error trying to determine if Manager lock is held", e);
+      }
+
+      LOG.info("Checking for existing fate transactions");
+      try {
+        // Adapted from UpgradeCoordinator.abortIfFateTransactions
+        if (!zoo.getChildren(Constants.ZFATE).isEmpty()) {
+          throw new IllegalStateException("Cannot complete upgrade preparation"
+              + " because FATE transactions exist. You can start a tserver, but"
+              + " not the Manager, then use the shell to delete completed"
+              + " transactions and fail pending or in-progress transactions."
+              + " Once all of the FATE transactions have been removed you can"
+              + " retry this operation.");
+        }
+      } catch (KeeperException | InterruptedException e) {
+        throw new IllegalStateException("Error checking for existing FATE transactions", e);
+      }
```

Comment:
Can we call out to FateAdmin for this check? Might be nice to delegate these responsibilities.

Reply:
I think I looked at that and it's meant to be a user-facing utility that has a text response. I'll look again to be sure.

Reply:
So, `FateAdmin` is in 1.10, but not 2.1. `FateCommand` in the shell exists in 2.1, but there is a deprecation warning to use `accumulo admin fate` instead. `accumulo admin fate --print` and `accumulo admin --summary` print formatted text to System.out. `accumulo admin fate --summary --printJson` prints json to System.out.

Reply:
Given the information above, I don't think we should delegate to a utility in 2.1.0. It might be more appropriate in 4.0 though given that solely looking in ZooKeeper for fate transactions will be insufficient.